pep/004-labels-and-namespaces.md

```diff
@@ -0,0 +1,386 @@
+
+If accepted labels would be stored in the bundle.json like so:
+
+```json
+{
+  "name": "mybun",
+  "version": "v0.1.0",
+  "labels": {
+    "app": "myapp",
+    "io.cnab/appVersion": "v1.2.3",
+    "internal": "true"
+  }
+}
+```
+
+[cnab-labels-proposal]: https://github.com/cnabio/cnab-spec/pull/409
+
+#### Installation Labels
+
+When a bundle is installed, the labels defined on the bundle are copied and used as labels on the installation.
+Users can apply additional labels using the --label flag, which can be repeated, has a short flag of `-l` and uses the following format: KEY=VALUE.
+
+Labels can be applied to an installation when the bundle is installed:
+```
+porter install wordpress --reference getporter/wordpress:v1.0 --label owner=carolyn
+```
+
+Labels can be applied after an installation is created using the `porter installation label` command:
+
+```
+$ porter installation label set wordpress --label env=dev --label internal=true
+# add (or updates) env=dev and internal=true to the wordpress installation
+
+$ porter installation label delete wordpress --label env
+# Remove the env label from the wordpress installation
+```
+
+🚨 The CNAB Spec does not cover storing installation data, and we should push for having a document representation of installations be added to the spec.
+
+```json
+{
+  "name": "wordpress",
+  "created": "TIMESTAMP",
+  "bundleRepository": "getporter/wordpress",
+  "version": "0.1.0",
+  "labels": {
+    "app": "wordpress",
+    "io.cnab/appVersion": "v1.2.3",
+    "owner": "carolyn"
+  },
+  "custom": {}
+}
+```
+
+Users can filter by labels when listing installations:
+
+```console
+$ porter installations list --label owner=carolyn
+# Filter installations by the label owner=carolyn
+```
+
+#### Claim Labels
+
+In Porter a claim is represented as a "run".
+Each time porter install/upgrade/invoke/uninstall is executed, that generates a claim (except when the action is stateless such as io.cnab.status).
+The user-defined labels provided when the bundle is installed are applied to both the installation and the claim.
+
+```
+porter install wordpress --reference getporter/wordpress:v1.0 --label owner=carolyn
+```
+
+**claim**
+```json
+{
+  "id": "CLAIM_ID",
+  "installation": "wordpress",
+  "action": "install",
+  "labels": {
+    "owner": "carolyn"
+  }
+}
+```
+
+During upgrade, invoke and uninstall the user-defined labels are applied to the claim only:
+
+```
+porter upgrade wordpress --label carolyn-test=true
+```
+
+**claim**
+```json
+{
+  "id": "CLAIM_ID",
+  "installation": "wordpress",
+  "action": "upgrade",
+  "labels": {
+    "carolyn-test": "true"
+  }
+}
+```
+
+#### Result Labels
+
+The spec allows for labels to be set on a result however we will not expose that in Porter.
+It is essentially reserved for custom tools and future use in Porter.
+
+#### Output Labels
+
+The spec allows for labels to be set on a result however we will not expose that in Porter.
```

Comment:
These two lines appear to be duplicates of the lines under `Result Labels` above.  Intended to repeat?

Reply:
copy/paste strikes again. I'll make it unique and cleanup.